File: csharp/binary-search/BinarySearch.cs
using System;

public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        int upper = input.Length;
        int lower = 0;
        int middle = upper / 2;
        while(lower < upper)
        {
            middle = lower + (upper - lower) / 2;
            if (input[middle] == value)
            {
                return middle;
            }
            else if (input[middle] > value)
            {
                upper = middle;
            }
            else
            {
                lower = middle + 1;
            }
        }
        return -1;
    }
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch


let find input value =
    let rec needle bottom hay =
        if Array.isEmpty hay then
            None
        else
            let midheight = (Array.length hay) >>> 1
            let bite = hay[midheight]

            if bite = value then Some(bottom + midheight)
            elif bite > value then needle bottom hay[.. (midheight - 1)]
            else needle (bottom + midheight + 1) hay[(midheight + 1) ..]

    needle 0 input
